app/src/lib/utils/filters.ts

```diff
@@ -1,4 +1,21 @@
+import { isStr } from '$lib/utils/string';
+
 // If a value occurs > 1 times then all but one will fail this condition.
 export function unique(value: any, index: number, array: any[]) {
 	return array.indexOf(value) === index;
 }
+
+/**
+ * Filters an array of objects based on a specific string value in a given key.
+ */
+export function filterStringByKey<T, K extends keyof T>(arr: T[], key: K, value: string): T[] {
```

Comment:
I would again drop the comment since the function name is sufficient to understand what's going on. But, could we still rename the function? It looks like it would be more appropriately called `filterObjectsByKey`?

Reply:
At the same time, I'm curious of the value of having this function at all given that it can be rather clearly as:

```ts
declare const items: ({ foo: number, bar: string })[]
const needle = "asfd"

items.filter(({ bar }) => bar.includes(needle))
```

Reply:
To me writing the filter statement is clearer than calling a function called `filterObjectsBySpecifyPropertyThatIncludesString(objects, "key", "asfd")`

Reply:
I don't have strong feelings about either approach :).

The main 2 reasons for which I opted for a separate dedicated function:
- I think inlining functional array methods into certain expressions can be actually a bit harder to read. In this case we'd write that into the ternary operator:

```ts
	$: filteredItems =
		filterText === undefined ? items : items.filter((it) => it[labelId].includes(filterText));
```
vs

```ts
$: filteredItems =
		filterText === undefined ? items : filterObjectsByKey(items, labelId, filterText);
```

- Also, just because `for`-loops are [generally more performant](https://medium.com/@maxheadway/exploring-performance-traditional-loops-vs-functional-methods-in-javascript-76dedc749375) than functional array methods (which kind of suck because they are more elegant in most cases). In the case that the list of e.g. target branches is huge, filtering the list will become more expensive and we should use as much performance optimizations as we can.

In practice that might not be the case most times so if you still prefer the functional method I'm ok with thah 👍

Reply:
Also, sorry if I keep pushing back on some of the comments. I appreciate both of you being so patient 😄😄

Reply:
> Also, sorry if I keep pushing back on some of the comments. I appreciate both of you being so patient 😄😄

I'm convinced that having discussion between multiple developers always ends up producing a much better result. Its great that you're pushing back and helping us consider the problem from your perspective. Please don't stop 😂!!

---

> Also, just because for-loops are [generally more performant](https://medium.com/@maxheadway/exploring-performance-traditional-loops-vs-functional-methods-in-javascript-76dedc749375) than functional array methods (which kind of sucks because they are more elegant in most cases). In the case that the list of e.g. target branches is huge, filtering the list will become more expensive and we should use as much performance optimizations as we can.

This is undoubtably true, but in the vast majority of situations the difference in performance is simply not relevant, even on lists of 10,000 odd items.

I think our general consensus is that code clarity should take a priority over scraping every last nanosecond of performance.

> I think inlining functional array methods into certain expressions can be actually a bit harder to read. In this case we'd write that into the ternary operator:

I'm still struggling to be convinced about your `filterObjectsByKey` because the name doesn't tell me that you're doing a string `.includes` on it. If anything `filterObjectsByKey` tells me that you've got a list of arbitrarily shaped objects and you're filtering out the objects which have one particular key.

I do see that you're point about the expression getting quite dense, so how about we define a function local to the component and say:

```ts

function filterItems(items: Item[], filterText: string | undefined) {
  if (filterText) {
    return items.filter((it) => it[labelId].includes(filterText));
  } else {
    return items
  }
}

$: filteredItems = filterItems(items, filterText)
```

By keeping the definition local, we're also leaving a convenient place where would could pull in a fuzzy comparison operation in the future

Reply:
> I'm convinced that having discussion between multiple developers always ends up producing a much better result. Its great that you're pushing back and helping us consider the problem from your perspective. Please don't stop 😂!!
>
Haha I agree 😄

> By keeping the definition local, we're also leaving a convenient place where would could pull in a fuzzy comparison operation in the future
>
Seems like a good compromise. Will add it as you described :)